go/pvl/interp.go

```diff
@@ -811,39 +819,65 @@ func runSelectorJSONInner(g ProofContextExt, state ScriptState, object *jsonw.Wr
 		"Selector entry not recognized: %v", selector)
 }
 
-// Take a template, substitute variables, and build the Regexp.
-func interpretRegex(g ProofContextExt, template string, state ScriptState) (*regexp.Regexp, libkb.ProofError) {
-	var perr libkb.ProofError = libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
-		"Could not build regex %v", template)
+func extractRegexDescriptor(g ProofContextExt, state ScriptState, obj *jsonw.Wrapper, templatekey string) (RegexDescriptor, libkb.ProofError) {
+	var desc RegexDescriptor
 
-	// Parse out bookends (^$) and option letters.
-	if !strings.HasPrefix(template, "^") {
-		return nil, perr
+	template, err := obj.AtKey(templatekey).GetString()
+	if err != nil {
+		debugWithState(g, state, "Could not get regex template: %v", err)
+		return RegexDescriptor{}, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get regex template")
 	}
-	lastDollar := strings.LastIndex(template, "$")
-	if lastDollar == -1 {
-		return nil, perr
+	desc.Template = template
+
+	if !obj.AtKey(RegexKeyCaseInsensitive).IsNil() {
+		caseinsensitive, err := obj.AtKey(RegexKeyCaseInsensitive).GetBool()
+		if err != nil {
+			return RegexDescriptor{}, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				"Regex option '%v' must be a bool", RegexKeyCaseInsensitive)
+		}
+		desc.CaseInsensitive = caseinsensitive
 	}
-	opts := template[lastDollar+1:]
-	if !regexp.MustCompile("[imsU]*").MatchString(opts) {
-		return nil, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
-			"Could not build regex: %v (%v)", "invalid options", template)
+
+	if !obj.AtKey(RegexKeyMultiLine).IsNil() {
+		multiline, err := obj.AtKey(RegexKeyMultiLine).GetBool()
+		if err != nil {
+			return RegexDescriptor{}, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				"Regex option '%v' must be a bool", RegexKeyMultiLine)
+		}
+		desc.MultiLine = multiline
 	}
-	var prefix = ""
-	optmap := make(map[rune]bool)
-	for _, opt := range opts {
-		optmap[opt] = true
+
+	return desc, nil
+}
+
+// Take a regex descriptor, do variable substitution, and build a regex.
+func interpretRegex(g ProofContextExt, state ScriptState, rdesc RegexDescriptor) (*regexp.Regexp, libkb.ProofError) {
+	// Check that regex is bounded by "^" and "$".
+	if !strings.HasPrefix(rdesc.Template, "^") {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not build regex: %v (%v)", "must start with '^'", rdesc.Template)
 	}
-	if len(opts) != len(optmap) {
+	if !strings.HasSuffix(rdesc.Template, "$") {
 		return nil, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
-			"Could not build regex: %v (%v)", "duplicate options", template)
+			"Could not build regex: %v (%v)", "must end with '$'", rdesc.Template)
+	}
```

Comment:
Right now `extractRegexDescriptor` and `interpretRegex` are always called one after the other. They are only separate so that error logging can see the `rdesc`. Yes it would be better to check earlier. But right now there's not really a parse vs run distinction other than being valid json. There's `validateChunk` but it is only run by `CheckProof`. I think the testing we do before deploying an new revision will be sure to hit any invalidities.